annotation/consume_trigger.go

```diff
@@ -71,9 +77,63 @@ type Prestring interface {
 	String() string
 }
 
+// Assignment is a struct that represents an assignment to an expression
+type Assignment struct {
+	LHSExprStr string
+	RHSExprStr string
+	Position   token.Position
+}
+
+func (a *Assignment) String() string {
+	return fmt.Sprintf("`%s` to `%s` at %s", a.RHSExprStr, a.LHSExprStr, a.Position)
+}
+
+// assignmentFlow is a struct that represents a flow of assignments
+type assignmentFlow struct {
+	assignments []Assignment
+}
+
+func (a *assignmentFlow) addEntry(entry Assignment) {
+	a.assignments = append(a.assignments, entry)
+}
+
+func (a *assignmentFlow) String() string {
+	if len(a.assignments) == 0 {
+		return ""
+	}
+
+	// fix point convergence in backprop could mean duplicate entries in this slice, hence we filter out the duplicates here
+	entries := make([]Assignment, 0, len(a.assignments))
+	seen := make(map[string]bool)
+	for _, entry := range a.assignments {
+		if !seen[entry.String()] {
+			seen[entry.String()] = true
+			entries = append(entries, entry)
+		}
+	}
+
+	// backprop algorithm populates assignment entries in backward order. Reverse entries to get forward order of assignments.
+	for i, j := 0, len(entries)-1; i < j; i, j = i+1, j-1 {
+		entries[i], entries[j] = entries[j], entries[i]
+	}
+
+	// build string slice
+	strs := make([]string, len(entries))
+	for i, entry := range entries {
+		strs[i] = entry.String()
+	}
+
+	// build the informative print string tracking the assignments
+	var sb strings.Builder
+	sb.WriteString(" via the assignment(s):\n\t\t-> ")
+	sb.WriteString(strings.Join(strs, ",\n\t\t-> "))
+	return sb.String()
+}
+
 // TriggerIfNonNil is triggered if the contained Annotation is non-nil
 type TriggerIfNonNil struct {
 	Ann Key
+	assignmentFlow
```

Comment:
Now, with using `orderedmap` for `assignmentFlow`, we should consider it in `copy`. Added code for it.

For `equals` we should not account for it. Consider the situation of multiple nilable flows reach a dereference site. We keep NilAway practical by tracking only one representative nil flow, and printing only one error message for the dereference site with one representative flow printed in the error message. However, considering `assignmentFlow` in `equals` would mean tracking all nilable flows in separate full triggers and reporting errors through all paths. This would likely incur a performance penalty, which we wouldn't want. I have added a note at the definition of `assignmentFlow` explaining this.